Adds PerformanceCurveBundle

trant22t opened the pull request:
This class allows users to store and plot multiple performance curves (same data / same metric / different models). 

Let me know what you think about the implementation. I will start writing tests when both of us are on the same page for the design.

trant22t committed:
Adds PerformanceCurveBundle, which allows users to store and plot multiple performance curves (same data / same metric / different models)

erp12 commented on performance_curves/performance_curve_bundle.py:
```diff
@@ -0,0 +1,54 @@
+                 y_true: np.ndarray,
+                 y_scores: List[np.ndarray],
+                 metric: Metric,
+                 model_names: List[str],
```
If we expect a 1:1 relationship between `y_scores` and `model_name` then we should consider asking for a `Mapping[src, np.ndarray]`. Unless order matters, in which case we can think of something else.

trant22t replied:
Good point! Both of these arguments are now condensed into one, named `method_scores` of type `Mapping[str, np.ndarray]`. I named them `method_` instead of `model_` following your first suggestion above.

erp12 commented on performance_curves/performance_curve_bundle.py:
```diff
@@ -0,0 +1,54 @@
+            all_case_counts.append(individual_curve.case_counts)
+        self.bundle = dict(zip(self.model_names, zip(all_performance_values, all_case_counts)))
+
+        if random_model:
+            random_curve = RandomPerformanceCurve(self.y_true, self.metric, num_trials, self.num_bins, random_seed)
+            self.bundle["random_model"] = tuple(zip(random_curve.performance_values, random_curve.case_counts))
+        if perfect_model:
+            perfect_curve = PerfectPerformanceCurve(self.y_true, self.metric, self.num_bins)
+            self.bundle["perfect_model"] = tuple(zip(perfect_curve.performance_values, perfect_curve.case_counts))
```
It seems likely that the random and perfect curves would always be handled as a special case. What do you think about putting them in their own attributes (not the `bundle`/`curves` dict) so that we don't need repeatedly check the key to make sure it isn't the random or perfect.

____

Separately, if we keep the dictionary then we should try to make keys we control ("random_model", "perfect_model") a bit more unlikely to conflict with the names the user passed in. I would add  one or two `_` to the beginning of the names we set.

trant22t replied:
Making `random_curve` and `perfect_curve` as part of the `PerformanceCurveBundle` class, but not part of the `curves` dictionary (attribute of `PerformanceCurveBundle`) indeed makes the code much cleaner! I confused myself at first by naming the attribute `bundle`, but renaming it makes the concept clearer. 

Cause we now separate `random_curve` and `perfect_curve` from the `curves` attribute, we don't need to worry about potential name conflicts with what the user passes in anymore.

erp12 commented on performance_curves/performance_curve_bundle.py:
```diff
@@ -0,0 +1,54 @@
+                 num_bins: Optional[int] = None,
+                 order_descending: bool = True,
+                 random_model: bool = True,
+                 random_seed: Optional[int] = None,
```
Users can set the seed in their own code. I don't think there is much value in controlling it ourselves.

trant22t replied:
We discussed this offline, but wanted to keep a record here just in case. It seems like `scikit-learn` uses `np.random.seed()` generator and ignores seed set by the built-in `random.seed()` ([reference](https://stackoverflow.com/questions/31057197/should-i-use-random-seed-or-numpy-random-seed-to-control-random-number-gener)). I removed the `random_seed` argument in our `RandomPerformanceCurve`, but it might be worth it to leave a note in our document saying that users should set seed using `numpy` instead of built-in `random`.

erp12 reviewed (changes requested):
This is a great start. I have a few comments in the code and two general points to think about.

- Do we want to limit the package to only be about models? We use that term a lot in this code, but we know that performance curves can be built from any ranking method. Is there an easy way we can avoid using "model" in our variables/attributes?

- There are many ways in which one could create a bundle. In this code, we create it from scores, but we could also create it from features + models. I think there is opportunity here in an abstract base class that defines the minimum contract of a performance curve bundle. Then different sub-classes could implement different ways of constructing bundles. What do you think?

erp12 commented on performance_curves/performance_curve_bundle.py:
```diff
@@ -0,0 +1,54 @@
+                                                order_descending)
+            all_performance_values.append(individual_curve.performance_values)
+            all_case_counts.append(individual_curve.case_counts)
+        self.bundle = dict(zip(self.model_names, zip(all_performance_values, all_case_counts)))
```
Naming suggestion: This entire class is the bundle so this attribute should be called something else. Perhaps "curves"?

This way, the user's code won't read like `bundle.bundle["whaterver"]` which is confusing.

trant22t replied:
Done!

trant22t committed:
Addresses naming issues, cleans up code and adds test cases for PerformanceCurveBundle

trant22t commented:
Regarding your two general points:  
- I renamed `model` with `method` throughout our code. Good call! 
- One challenge about allowing users to supply `features + models` is that it's hard for us to decide how a model would look like. Is it going to be a `scikit-learn` `estimator`? Is it required to have the `predict` method? Could it just be a `transformer`?  Regardless, coming up with a clear definition for `model` seems out of scope for our package (at least at this point). I'm good with the current design of `PerformanceCurveBundle` as an MVP - we could revisit this issue as the package has more users and requires greater flexibility. 

I resolved all the other issues according to your suggestions. All very good points! Thank you for pointing them out! I also added the test cases in this change as it seems like we're both generally happy with the implementation now. Let me know what you think.

erp12 commented on performance_curves/performance_curve_bundle.py:
```diff
@@ -0,0 +1,48 @@
+                 order_descending: bool = True,
+                 make_perfect: bool = True,
+                 make_random: bool = True,
+                 num_trials: Optional[int] = 1):
```
It doesn't seem like `num_trials` can be `None`. Is that correct? If so, then it is not `Optional`.

trant22t replied:
Good catch! `num_trials` is changed to a non-optional argument.

erp12 reviewed (approved):
Looks great to me! I have one tiny comment about a type annotation.  After that is fixed, feel free to squash and merge.

trant22t committed:
Change num_trials argument to be not optional
